File: curriculum/12_distributed_job_scheduler/node-impl/src/scheduler.ts
export enum Priority {
  Low = 'low',
  Normal = 'normal',
  High = 'high',
}

export enum JobStatus {
  Pending = 'pending',
  Running = 'running',
  Completed = 'completed',
  Failed = 'failed',
  Cancelled = 'cancelled',
}

export enum JobResult {
  Completed = 'completed',
  Failed = 'failed',
  Cancelled = 'cancelled',
}

export type JobRequest = {
  name: string;
  payload?: Record<string, string>;
  interval?: string;
  priority?: Priority;
  dependencies?: string[];
  maxAttempts?: number;
  initialBackoffMs?: number;
  runAfterMs?: number;
};

export type Job = {
  id: string;
  name: string;
  payload: Record<string, string>;
  status: JobStatus;
  priority: Priority;
  dependencies: string[];
  intervalMs: number | null;
  dueAtMs: number;
  createdAtMs: number;
  attempt: number;
  maxAttempts: number;
  initialBackoffMs: number;
  lastError: string | null;
  cancelReason: string | null;
};

export type Lock = {
  resourceId: string;
  ownerNodeId: string;
  ownerWorkerId: string;
  fencingToken: number;
  leaseExpiresAtMs: number;
};

export type Health = {
  nodeId: string;
  leaderId: string | null;
  queueDepth: number;
  runningJobs: number;
  expiredLocks: number;
};

type Now = () => number;

export class Election {
  private leaderId: string | null = null;
  private leaseExpiresAtMs = 0;

  constructor(
    private readonly now: Now,
    private readonly leaseMs: number,
  ) {}

  elect(processIds: readonly string[]): string | null {
    const leader = [...processIds].sort().at(-1) ?? null;
    this.leaderId = leader;
    this.leaseExpiresAtMs = this.now() + this.leaseMs;
    return leader;
  }

  isLeader(processId: string): boolean {
    return this.leaderId === processId && this.now() < this.leaseExpiresAtMs;
  }

  currentLeader(): string | null {
    return this.leaderId;
  }
}

export class LockManager {
  private readonly locks = new Map<string, Lock>();
  private readonly tokens = new Map<string, number>();

  constructor(private readonly now: Now) {}

  acquire(resourceId: string, ownerNodeId: string, ownerWorkerId: string, leaseMs: number): Lock {
    const current = this.locks.get(resourceId);
    if (current !== undefined && this.now() < current.leaseExpiresAtMs) {
      throw new Error(`resource ${resourceId} is locked`);
    }
    const fencingToken = (this.tokens.get(resourceId) ?? 0) + 1;
    this.tokens.set(resourceId, fencingToken);
    const lock: Lock = {
      resourceId,
      ownerNodeId,
      ownerWorkerId,
      fencingToken,
      leaseExpiresAtMs: this.now() + leaseMs,
    };
    this.locks.set(resourceId, lock);
    return lock;
  }

  validate(resourceId: string, fencingToken: number): void {
    const lock = this.locks.get(resourceId);
    if (lock === undefined) {
      throw new Error('lock not found');
    }
    if (this.now() > lock.leaseExpiresAtMs) {
      throw new Error('lock expired');
    }
    if (lock.fencingToken !== fencingToken) {
      throw new Error('stale fencing token');
    }
  }

  release(resourceId: string): void {
    this.locks.delete(resourceId);
  }

  expiredCount(): number {
    let count = 0;
    for (const lock of this.locks.values()) {
      if (this.now() > lock.leaseExpiresAtMs) {
        count += 1;
      }
    }
    return count;
  }
}

export class Scheduler {
  private readonly jobs = new Map<string, Job>();
  private readonly locks: LockManager;
  private election: Election;
  private sequence = 0;

  constructor(
    private readonly nodeId: string,
    private readonly now: Now = () => Date.now(),
  ) {
    this.locks = new LockManager(now);
    this.election = new Election(now, 5_000);
  }

  submit(request: JobRequest): Job {
    if (request.name.trim().length === 0) {
      throw new Error('job name is required');
    }
    const intervalMs = request.interval === undefined ? null : parseDurationMs(request.interval);
    const dependencies = [...(request.dependencies ?? [])];
    for (const dependency of dependencies) {
      if (!this.jobs.has(dependency)) {
        throw new Error(`dependency ${dependency} does not exist`);
      }
    }

    this.sequence += 1;
    const job: Job = {
      id: `job-${this.sequence.toString().padStart(6, '0')}`,
      name: request.name,
      payload: { ...(request.payload ?? {}) },
      status: JobStatus.Pending,
      priority: request.priority ?? Priority.Normal,
      dependencies,
      intervalMs,
      dueAtMs: this.now() + (request.runAfterMs ?? 0),
      createdAtMs: this.now(),
      attempt: 0,
      maxAttempts: Math.max(1, request.maxAttempts ?? 1),
      initialBackoffMs: Math.max(1, request.initialBackoffMs ?? 1_000),
      lastError: null,
      cancelReason: null,
    };
    this.jobs.set(job.id, job);
    logJson('job_submitted', { jobId: job.id, priority: job.priority, dependencies: dependencies.length });
    return cloneJob(job);
  }

  becomeLeader(peers: readonly string[], leaseMs: number): void {
    this.election = new Election(this.now, leaseMs);
    this.election.elect(peers);
  }

  dispatchNext(workerId: string, leaseMs: number): { job: Job; lock: Lock } {
    const candidates = [...this.jobs.values()]
      .filter((job) => job.status === JobStatus.Pending && this.dependenciesCompleted(job))
      .sort(compareJobs);

    for (const candidate of candidates) {
      try {
        return this.dispatchJob(candidate.id, workerId, leaseMs);
      } catch (error) {
        if (!(error instanceof Error)) {
          throw error;
        }
      }
    }
    throw new Error('no dispatchable job');
  }

  dispatchJob(jobId: string, workerId: string, leaseMs: number): { job: Job; lock: Lock } {
    if (!this.election.isLeader(this.nodeId)) {
      throw new Error('not leader');
    }
    const job = this.requireJob(jobId);
    if (job.status !== JobStatus.Pending) {
      throw new Error(`job is ${job.status}`);
    }
    if (!this.dependenciesCompleted(job)) {
      throw new Error('dependencies not completed');
    }

    const lock = this.locks.acquire(job.id, this.nodeId, workerId, leaseMs);
    job.status = JobStatus.Running;
    job.attempt += 1;
    this.jobs.set(job.id, job);
    logJson('job_dispatched', { jobId: job.id, workerId, fencingToken: lock.fencingToken });
    return { job: cloneJob(job), lock: { ...lock } };
  }

  complete(jobId: string, fencingToken: number, result: JobResult, message: string | null = null): void {
    this.locks.validate(jobId, fencingToken);
    const job = this.requireJob(jobId);
    switch (result) {
      case JobResult.Completed:
        job.status = JobStatus.Completed;
        break;
      case JobResult.Cancelled:
        job.status = JobStatus.Cancelled;
        break;
      case JobResult.Failed:
        job.lastError = message;
        if (job.attempt < job.maxAttempts) {
          job.status = JobStatus.Pending;
          job.dueAtMs = this.now() + retryBackoffMs(job.initialBackoffMs, job.attempt);
        } else {
          job.status = JobStatus.Failed;
        }
        break;
    }
    this.jobs.set(job.id, job);
    this.locks.release(job.id);
    logJson('job_completed', { jobId: job.id, status: job.status, attempt: job.attempt });
  }

  cancel(jobId: string, reason: string): void {
    const job = this.requireJob(jobId);
    if ([JobStatus.Completed, JobStatus.Failed, JobStatus.Cancelled].includes(job.status)) {
      throw new Error(`terminal job state ${job.status}`);
    }
    job.status = JobStatus.Cancelled;
    job.cancelReason = reason;
    this.jobs.set(job.id, job);
    this.locks.release(job.id);
    logJson('job_cancelled', { jobId: job.id, reason });
  }

  getJob(jobId: string): Job {
    return cloneJob(this.requireJob(jobId));
  }

  health(): Health {
    const jobs = [...this.jobs.values()];
    return {
      nodeId: this.nodeId,
      leaderId: this.election.currentLeader(),
      queueDepth: jobs.filter((job) => job.status === JobStatus.Pending && this.dependenciesCompleted(job)).length,
      runningJobs: jobs.filter((job) => job.status === JobStatus.Running).length,
      expiredLocks: this.locks.expiredCount(),
    };
  }

  private requireJob(jobId: string): Job {
    const job = this.jobs.get(jobId);
    if (job === undefined) {
      throw new Error('job not found');
    }
    return { ...job, dependencies: [...job.dependencies], payload: { ...job.payload } };
  }

  private dependenciesCompleted(job: Job): boolean {
    return job.dependencies.every((parentId) => this.jobs.get(parentId)?.status === JobStatus.Completed);
  }
}

function parseDurationMs(input: string): number {
  if (input.endsWith('s')) {
    return parsePositiveInteger(input.slice(0, -1), input) * 1_000;
  }
  if (input.endsWith('m')) {
    return parsePositiveInteger(input.slice(0, -1), input) * 60_000;
  }
  throw new Error(`invalid interval ${input}`);
}

function parsePositiveInteger(raw: string, original: string): number {
  if (!/^\d+$/.test(raw)) {
    throw new Error(`invalid interval ${original}`);
  }
  const value = Number(raw);
  if (!Number.isSafeInteger(value) || value <= 0) {
    throw new Error(`invalid interval ${original}`);
  }
  return value;
}

function compareJobs(left: Job, right: Job): number {
  const priorityDelta = priorityRank(right.priority) - priorityRank(left.priority);
  if (priorityDelta !== 0) {
    return priorityDelta;
  }
  const dueDelta = left.dueAtMs - right.dueAtMs;
  if (dueDelta !== 0) {
    return dueDelta;
  }
  return left.createdAtMs - right.createdAtMs;
}

function priorityRank(priority: Priority): number {
  switch (priority) {
    case Priority.High:
      return 3;
    case Priority.Normal:
      return 2;
    case Priority.Low:
      return 1;
  }
}

function retryBackoffMs(initialBackoffMs: number, attempt: number): number {
  return initialBackoffMs * 2 ** Math.max(0, attempt - 1);
}

function cloneJob(job: Job): Job {
  return {
    ...job,
    dependencies: [...job.dependencies],
    payload: { ...job.payload },
  };
}

export function logJson(event: string, fields: Record<string, string | number | boolean>): void {
  console.log(JSON.stringify({ event, ...fields, timestamp: new Date().toISOString() }));
}

File: curriculum/12_distributed_job_scheduler/go-impl/internal/scheduler/scheduler.go
package scheduler

import (
	"context"
	"errors"
	"fmt"
	"log/slog"
	"sort"
	"strings"
	"sync"
	"time"
)

type Clock struct {
	Now func() time.Time
}

func (c Clock) now() time.Time {
	if c.Now == nil {
		return time.Now().UTC()
	}
	return c.Now().UTC()
}

type Priority string

const (
	PriorityHigh   Priority = "high"
	PriorityNormal Priority = "normal"
	PriorityLow    Priority = "low"
)

type Status string

const (
	StatusPending   Status = "pending"
	StatusRunning   Status = "running"
	StatusCompleted Status = "completed"
	StatusFailed    Status = "failed"
	StatusCancelled Status = "cancelled"
)

type Result string

const (
	ResultCompleted Result = "completed"
	ResultFailed    Result = "failed"
	ResultCancelled Result = "cancelled"
)

type JobRequest struct {
	Name           string
	Payload        map[string]string
	Interval       string
	Priority       Priority
	Dependencies   []string
	MaxAttempts    int
	InitialBackoff time.Duration
	RunAfter       time.Duration
}

type Job struct {
	ID             string            `json:"id"`
	Name           string            `json:"name"`
	Payload        map[string]string `json:"payload,omitempty"`
	Status         Status            `json:"status"`
	Priority       Priority          `json:"priority"`
	Dependencies   []string          `json:"dependencies"`
	Interval       time.Duration     `json:"interval"`
	DueAt          time.Time         `json:"due_at"`
	CreatedAt      time.Time         `json:"created_at"`
	UpdatedAt      time.Time         `json:"updated_at"`
	Attempt        int               `json:"attempt"`
	MaxAttempts    int               `json:"max_attempts"`
	InitialBackoff time.Duration     `json:"initial_backoff"`
	LastError      string            `json:"last_error,omitempty"`
	CancelReason   string            `json:"cancel_reason,omitempty"`
}

type Lock struct {
	ResourceID     string    `json:"resource_id"`
	OwnerNodeID    string    `json:"owner_node_id"`
	OwnerWorkerID  string    `json:"owner_worker_id"`
	FencingToken   int64     `json:"fencing_token"`
	LeaseExpiresAt time.Time `json:"lease_expires_at"`
}

type Health struct {
	NodeID       string `json:"node_id"`
	LeaderID     string `json:"leader_id"`
	QueueDepth   int    `json:"queue_depth"`
	RunningJobs  int    `json:"running_jobs"`
	ExpiredLocks int    `json:"expired_locks"`
}

type Scheduler struct {
	mu       sync.Mutex
	nodeID   string
	clock    Clock
	jobs     map[string]Job
	locks    *LockManager
	election *Election
	seq      int
}

func NewScheduler(nodeID string, clock Clock) *Scheduler {
	return &Scheduler{
		nodeID:   nodeID,
		clock:    clock,
		jobs:     map[string]Job{},
		locks:    NewLockManager(clock),
		election: NewElection(clock, 5*time.Second),
	}
}

func (s *Scheduler) Submit(_ context.Context, req JobRequest) (Job, error) {
	s.mu.Lock()
	defer s.mu.Unlock()

	if strings.TrimSpace(req.Name) == "" {
		return Job{}, errors.New("job name is required")
	}
	interval := time.Duration(0)
	if req.Interval != "" {
		parsed, err := time.ParseDuration(req.Interval)
		if err != nil || parsed <= 0 {
			return Job{}, fmt.Errorf("invalid interval %q", req.Interval)
		}
		interval = parsed
	}
	priority := req.Priority
	if priority == "" {
		priority = PriorityNormal
	}
	if priorityRank(priority) == 0 {
		return Job{}, fmt.Errorf("invalid priority %q", priority)
	}
	for _, parentID := range req.Dependencies {
		_, ok := s.jobs[parentID]
		if !ok {
			return Job{}, fmt.Errorf("dependency %s does not exist", parentID)
		}
		if parentID == "" {
			return Job{}, errors.New("empty dependency id")
		}
	}
	now := s.clock.now()
	s.seq++
	maxAttempts := req.MaxAttempts
	if maxAttempts <= 0 {
		maxAttempts = 1
	}
	backoff := req.InitialBackoff
	if backoff <= 0 {
		backoff = time.Second
	}
	job := Job{
		ID:             fmt.Sprintf("job-%06d", s.seq),
		Name:           req.Name,
		Payload:        req.Payload,
		Status:         StatusPending,
		Priority:       priority,
		Dependencies:   append([]string(nil), req.Dependencies...),
		Interval:       interval,
		DueAt:          now.Add(req.RunAfter),
		CreatedAt:      now,
		UpdatedAt:      now,
		MaxAttempts:    maxAttempts,
		InitialBackoff: backoff,
	}
	s.jobs[job.ID] = job
	slog.Info("job_submitted", "job_id", job.ID, "priority", job.Priority, "dependencies", len(job.Dependencies))
	return job, nil
}

func (s *Scheduler) BecomeLeader(peers []string, lease time.Duration) {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.election.lease = lease
	s.election.Elect(peers)
}

func (s *Scheduler) DispatchNext(ctx context.Context, workerID string, lease time.Duration) (Job, Lock, error) {
	s.mu.Lock()
	candidates := make([]Job, 0, len(s.jobs))
	for _, job := range s.jobs {
		if job.Status == StatusPending && s.dependenciesCompletedLocked(job) {
			candidates = append(candidates, job)
		}
	}
	s.mu.Unlock()

	sortJobs(candidates)
	for _, job := range candidates {
		dispatched, lock, err := s.DispatchJob(ctx, job.ID, workerID, lease)
		if err == nil {
			return dispatched, lock, nil
		}
	}
	return Job{}, Lock{}, errors.New("no dispatchable job")
}

func (s *Scheduler) DispatchJob(_ context.Context, jobID string, workerID string, lease time.Duration) (Job, Lock, error) {
	s.mu.Lock()
	defer s.mu.Unlock()
	if !s.election.IsLeader(s.nodeID) {
		return Job{}, Lock{}, errors.New("not leader")
	}
	job, ok := s.jobs[jobID]
	if !ok {
		return Job{}, Lock{}, errors.New("job not found")
	}
	if job.Status != StatusPending {
		return Job{}, Lock{}, fmt.Errorf("job is %s", job.Status)
	}
	if !s.dependenciesCompletedLocked(job) {
		return Job{}, Lock{}, errors.New("dependencies not completed")
	}
	lock, err := s.locks.acquireLocked(jobID, s.nodeID, workerID, lease)
	if err != nil {
		return Job{}, Lock{}, err
	}
	job.Status = StatusRunning
	job.Attempt++
	job.UpdatedAt = s.clock.now()
	s.jobs[jobID] = job
	slog.Info("job_dispatched", "job_id", job.ID, "worker_id", workerID, "fencing_token", lock.FencingToken)
	return job, lock, nil
}

func (s *Scheduler) Complete(_ context.Context, jobID string, token int64, result Result, message string) error {
	s.mu.Lock()
	defer s.mu.Unlock()
	if err := s.locks.validateLocked(jobID, token); err != nil {
		return err
	}
	job, ok := s.jobs[jobID]
	if !ok {
		return errors.New("job not found")
	}
	now := s.clock.now()
	switch result {
	case ResultCompleted:
		job.Status = StatusCompleted
	case ResultCancelled:
		job.Status = StatusCancelled
	case ResultFailed:
		job.LastError = message
		if job.Attempt < job.MaxAttempts {
			job.Status = StatusPending
			job.DueAt = now.Add(backoff(job.InitialBackoff, job.Attempt))
		} else {
			job.Status = StatusFailed
		}
	default:
		return fmt.Errorf("unknown result %q", result)
	}
	job.UpdatedAt = now
	s.jobs[jobID] = job
	s.locks.releaseLocked(jobID)
	slog.Info("job_completed", "job_id", job.ID, "status", job.Status, "attempt", job.Attempt)
	return nil
}

func (s *Scheduler) Cancel(_ context.Context, jobID string, reason string) error {
	s.mu.Lock()
	defer s.mu.Unlock()
	job, ok := s.jobs[jobID]
	if !ok {
		return errors.New("job not found")
	}
	if job.Status == StatusCompleted || job.Status == StatusFailed || job.Status == StatusCancelled {
		return fmt.Errorf("terminal job state %s", job.Status)
	}
	job.Status = StatusCancelled
	job.CancelReason = reason
	job.UpdatedAt = s.clock.now()
	s.jobs[jobID] = job
	s.locks.releaseLocked(jobID)
	slog.Info("job_cancelled", "job_id", job.ID, "reason", reason)
	return nil
}

func (s *Scheduler) GetJob(jobID string) (Job, bool) {
	s.mu.Lock()
	defer s.mu.Unlock()
	job, ok := s.jobs[jobID]
	return job, ok
}

func (s *Scheduler) Health() Health {
	s.mu.Lock()
	defer s.mu.Unlock()
	health := Health{NodeID: s.nodeID, LeaderID: s.election.leaderID, ExpiredLocks: s.locks.expiredLocked()}
	for _, job := range s.jobs {
		switch job.Status {
		case StatusPending:
			if s.dependenciesCompletedLocked(job) {
				health.QueueDepth++
			}
		case StatusRunning:
			health.RunningJobs++
		}
	}
	return health
}

func (s *Scheduler) dependenciesCompletedLocked(job Job) bool {
	for _, parentID := range job.Dependencies {
		parent := s.jobs[parentID]
		if parent.Status != StatusCompleted {
			return false
		}
	}
	return true
}

type Election struct {
	clock       Clock
	lease       time.Duration
	leaderID    string
	leaseExpiry time.Time
}

func NewElection(clock Clock, lease time.Duration) *Election {
	return &Election{clock: clock, lease: lease}
}

func (e *Election) Elect(processIDs []string) string {
	if len(processIDs) == 0 {
		e.leaderID = ""
		return ""
	}
	leader := processIDs[0]
	for _, id := range processIDs[1:] {
		if id > leader {
			leader = id
		}
	}
	e.leaderID = leader
	e.leaseExpiry = e.clock.now().Add(e.lease)
	return leader
}

func (e *Election) IsLeader(processID string) bool {
	return e.leaderID == processID && e.clock.now().Before(e.leaseExpiry)
}

type LockManager struct {
	mu     sync.Mutex
	clock  Clock
	locks  map[string]Lock
	tokens map[string]int64
}

func NewLockManager(clock Clock) *LockManager {
	return &LockManager{clock: clock, locks: map[string]Lock{}, tokens: map[string]int64{}}
}

func (l *LockManager) Acquire(resourceID string, ownerNodeID string, ownerWorkerID string, lease time.Duration) (Lock, error) {
	l.mu.Lock()
	defer l.mu.Unlock()
	return l.acquireLocked(resourceID, ownerNodeID, ownerWorkerID, lease)
}

func (l *LockManager) Validate(resourceID string, token int64) error {
	l.mu.Lock()
	defer l.mu.Unlock()
	return l.validateLocked(resourceID, token)
}

func (l *LockManager) acquireLocked(resourceID string, ownerNodeID string, ownerWorkerID string, lease time.Duration) (Lock, error) {
	now := l.clock.now()
	if current, ok := l.locks[resourceID]; ok && now.Before(current.LeaseExpiresAt) {
		return Lock{}, fmt.Errorf("resource %s is locked", resourceID)
	}
	l.tokens[resourceID]++
	lock := Lock{ResourceID: resourceID, OwnerNodeID: ownerNodeID, OwnerWorkerID: ownerWorkerID, FencingToken: l.tokens[resourceID], LeaseExpiresAt: now.Add(lease)}
	l.locks[resourceID] = lock
	return lock, nil
}

func (l *LockManager) validateLocked(resourceID string, token int64) error {
	lock, ok := l.locks[resourceID]
	if !ok {
		return errors.New("lock not found")
	}
	if l.clock.now().After(lock.LeaseExpiresAt) {
		return errors.New("lock expired")
	}
	if token != lock.FencingToken {
		return errors.New("stale fencing token")
	}
	return nil
}

func (l *LockManager) releaseLocked(resourceID string) {
	delete(l.locks, resourceID)
}

func (l *LockManager) expiredLocked() int {
	now := l.clock.now()
	count := 0
	for _, lock := range l.locks {
		if now.After(lock.LeaseExpiresAt) {
			count++
		}
	}
	return count
}

func sortJobs(jobs []Job) {
	sort.SliceStable(jobs, func(i, j int) bool {
		left := jobs[i]
		right := jobs[j]
		if priorityRank(left.Priority) != priorityRank(right.Priority) {
			return priorityRank(left.Priority) > priorityRank(right.Priority)
		}
		if !left.DueAt.Equal(right.DueAt) {
			return left.DueAt.Before(right.DueAt)
		}
		return left.CreatedAt.Before(right.CreatedAt)
	})
}

func priorityRank(priority Priority) int {
	switch priority {
	case PriorityHigh:
		return 3
	case PriorityNormal:
		return 2
	case PriorityLow:
		return 1
	default:
		return 0
	}
}

func backoff(initial time.Duration, attempt int) time.Duration {
	if attempt <= 1 {
		return initial
	}
	return initial << (attempt - 1)
}
